File: Week 1 C/credit.c
#include <cs50.h>
#include <stdio.h>

bool CHECKSUM(long digits, long card_number)
{
long single_digit;
long checksum = 0;
while (card_number != 0)
{
    single_digit = card_number % 10;
	card_number = card_number / 10;
	// sum on odd digits.
	checksum += single_digit;
    single_digit = card_number % 10;
	card_number = card_number / 10;
	// time 2, then % and / , then add them all
	checksum += (single_digit * 2 / 10) + (single_digit * 2 % 10);
}
if (checksum % 10 == 0)
{
return true;
}
else
return false;
}

int main(void)
{
long number = get_long("Enter the number:");
// the length of the number
// f2d is the first 2-digit number
long f2d = number;
int length = 2;
while ( f2d > 100 )
{
	f2d = f2d / 10;
	length++;
}

if ( (length == 15) && (f2d == 34 || f2d == 37) && (CHECKSUM(length,number)) )
{
printf ("AMEX\n");
}
else if ( (length == 16) && (f2d > 50) && (f2d < 56) && (CHECKSUM(length,number)) )
{
printf("MASTERCARD\n");
}
else if ( ((length == 16) || (length == 13)) && (f2d / 10 == 4) && (CHECKSUM(length,number)))
{
printf("VISA\n");
}
else
{
printf("INVALID\n");
}
}

File: Week 6 Python/credit.py
# TODO
from cs50 import get_int
from cs50 import get_string

# Better to treat 'number' as 'string'
def main():
    while True:
        try:
            number = get_string("Number: ")
            if int(number):
                break
        except ValueError:
            continue
    if len(number) == 15 and valid(number) and (number[:2] in ["34", "37"]):
        print ("AMEX")
    elif len(number) == 16 and valid(number) and (number[:2] in ["51", "52", "53", "54", "55"]):
        print ("MASTERCARD")
    elif (len(number) in [13, 16]) and valid(number) and (number[:1] == "4"):
        print ("VISA")
    else:
        print ("INVALID")


def valid(number):
    number = number[::-1]
    sum = 0
    for i in range(0,len(number),2):
        if i == len(number) - 1:
            sum += int(number[i])
            break
        sum += (int(number[i])) + (((int(number[i + 1]) * 2) // 10)) + (int(number[i + 1]) * 2 % 10)
    if sum % 10 == 0:
        return True
    else:
        return False

main()